File: android-app/core-crypto/src/main/kotlin/ua/authapp/crypto/totp/Totp.kt
package ua.authapp.crypto.totp

import ua.authapp.crypto.mac.HashAlgorithm
import ua.authapp.crypto.mac.MacFactory
import ua.authapp.crypto.truncation.DynamicTruncation

/**
 * TOTP за RFC 6238: лічильник = floor((unixTime − T0) / period), T0 = 0.
 * Підтримує стандартні (6–8 цифр) та розширені (до 10 цифр) конфігурації.
 */
object Totp {

    fun generate(
        algorithm: HashAlgorithm,
        secret: ByteArray,
        unixTimeSeconds: Long,
        period: Int = 30,
        digits: Int = 6,
    ): String {
        require(period in 15..120) { "Період має бути від 15 до 120 секунд, отримано $period" }
        require(secret.isNotEmpty()) { "Секрет не може бути порожнім" }
        val counter = unixTimeSeconds / period
        val counterBytes = ByteArray(8)
        for (i in 0 until 8) {
            counterBytes[7 - i] = ((counter ushr (i * 8)) and 0xFF).toByte()
        }
        val mac = MacFactory.compute(algorithm, secret, counterBytes)
        return DynamicTruncation.truncate(algorithm, mac, digits).code
    }

    /**
     * Перевірка коду з вікном ±window періодів (стенд використовує window=1;
     * FR-022). Порівняння — константним часом.
     */
    fun verify(
        algorithm: HashAlgorithm,
        secret: ByteArray,
        code: String,
        unixTimeSeconds: Long,
        period: Int = 30,
        digits: Int = 6,
        window: Int = 1,
    ): Boolean {
        for (step in -window..window) {
            val candidate = generate(algorithm, secret, unixTimeSeconds + step.toLong() * period, period, digits)
            if (constantTimeEquals(candidate, code)) return true
        }
        return false
    }

    private fun constantTimeEquals(a: String, b: String): Boolean {
        if (a.length != b.length) return false
        var diff = 0
        for (i in a.indices) diff = diff or (a[i].code xor b[i].code)
        return diff == 0
    }
}

File: web-stand/src/crypto/totp.js
// TOTP за RFC 6238 — дзеркало Kotlin-ядра.
import { computeMac } from './mac.js';
import { truncate } from './truncation.js';

/**
 * @param {string} alg канонічний ідентифікатор алгоритму
 * @param {Uint8Array} secret
 * @param {number} timeSeconds unix-час у секундах
 * @param {number} period 15..120
 * @param {number} digits 6..10
 * @returns {string} код із провідними нулями
 */
export function generateTotp(alg, secret, timeSeconds, period = 30, digits = 6) {
  if (period < 15 || period > 120) {
    throw new Error(`Період має бути від 15 до 120 секунд, отримано ${period}`);
  }
  if (!secret || secret.length === 0) throw new Error('Секрет не може бути порожнім');

  const counter = BigInt(Math.floor(timeSeconds / period));
  const counterBytes = new Uint8Array(8);
  let c = counter;
  for (let i = 7; i >= 0; i--) {
    counterBytes[i] = Number(c & 0xffn);
    c >>= 8n;
  }
  return truncate(alg, computeMac(alg, secret, counterBytes), digits).code;
}

/**
 * Валідація коду з вікном ±window періодів (FR-019/FR-022 специфікації:
 * стенд приймає коди в межах ±1 періоду для компенсації зсуву годинника).
 */
export function verifyTotp(alg, secret, code, timeSeconds, period = 30, digits = 6, window = 1) {
  const normalized = code.replace(/\s+/g, '');
  for (let step = -window; step <= window; step++) {
    if (generateTotp(alg, secret, timeSeconds + step * period, period, digits) === normalized) {
      return true;
    }
  }
  return false;
}
